File: lintcode/465.kth-smallest-sum-in-two-sorted-arrays.py
import heapq
class Solution:
    """
    @param A: an integer arrays sorted in ascending order
    @param B: an integer arrays sorted in ascending order
    @param k: An integer
    @return: An integer
    """
    def kthSmallestSum(self, A, B, k):
        # write your code here

        hq = []
        for i in range(len(B)):
            heapq.heappush(hq, (B[i] + A[0], i, 0))

        for i in range(k - 1):
            val, x, y = hq[0]
            heapq.heappop(hq)

            if y + 1 < len(A):
                heapq.heappush(hq, (B[x] + A[y + 1], x, y + 1))

        return hq[0][0]

File: lintcode/465.kth-smallest-sum-in-two-sorted-arrays.cpp
class Solution {
public:
    /**
     * @param A: an integer arrays sorted in ascending order
     * @param B: an integer arrays sorted in ascending order
     * @param k: An integer
     * @return: An integer
     */
    int kthSmallestSum(vector<int> &A, vector<int> &B, int k) {
        // write your code here
        
        struct Tuple{
            int x, y, val;
            Tuple(int _x, int _y, int _val): x(_x), y(_y), val(_val){}
        };
        
        auto cmp = [](const Tuple& left, const Tuple& right)
        {
            return left.val > right.val;
        };
        
        priority_queue<Tuple, vector<Tuple>, decltype(cmp)> pq(cmp);
        for (int i = 0; i < B.size(); ++i)
        {
            pq.push(Tuple(i, 0, B[i] + A[0]));
        }
        
        for (int i = 0; i < k - 1; ++i)
        {
            Tuple t = pq.top();
            pq.pop();
            
            if (t.y + 1 < A.size())
            {
                pq.push(Tuple(t.x, t.y + 1, B[t.x] + A[t.y + 1]));
            }
        }
        
        return pq.top().val;
    }
};